// @pulumi/pulumi/runtime/closure/v8.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
// This file provides a low-level interface to a few V8 runtime objects.
// We will use this low-level interface when serializing closures to walk the scope
// chain and find the value of free variables captured by closures, as well as getting
// source-level debug information so that we can present high-quality error messages.
//
// As a side-effect of importing this file, we must enable the --allow-natives-syntax V8
// flag. This is because we are using V8 intrinsics in order to implement this module.
const semver = require("semver");
const v8 = require("v8");
v8.setFlagsFromString("--allow-natives-syntax");
// We use four V8 intrinsics in this file. The first, `FunctionGetScript`, gets
// a `Script` object given a JavaScript function. The `Script` object contains metadata
// about the function's source definition.
const getScript = 
// The use of the Function constructor here and elsewhere in this file is because
// because V8 intrinsics are not valid JavaScript identifiers; they all begin with '%',
// which means that the TypeScript compiler issues errors for them.
new Function("func", "return %FunctionGetScript(func);");
// Check if we're on node 10.  We need to call specific intrinsics depending on which version we're on.
const isNodeV10 = semver.gte(process.version, "10.0.0");
// The second intrinsic is `FunctionGetScriptSourcePosition`, which does about what you'd
// expect. It returns a `V8SourcePosition`, which can be passed to `V8Script::locationFromPosition`
// to produce a `V8SourceLocation`.
const getSourcePosition = new Function("func", "return %FunctionGetScriptSourcePosition(func);");
const scriptPositionInfo = isNodeV10
    ? new Function("script", "pos", "return %ScriptPositionInfo(script, pos, false);")
    : undefined;
// The last two intrinsics are `GetFunctionScopeCount` and `GetFunctionScopeDetails`.
// The former function returns the number of scopes in a given function's scope chain, while
// the latter function returns the i'th entry in a function's scope chain, given a function and
// index i.
const getFunctionScopeDetails = new Function("func", "index", "return %GetFunctionScopeDetails(func, index);");
const getFunctionScopeCount = new Function("func", "return %GetFunctionScopeCount(func);");
// All of these functions contain syntax that is not legal TS/JS (i.e. "%Whatever").  As such,
// we cannot serialize them.  In case they somehow get captured, just block them from closure
// serialization entirely.
getScript.doNotCapture = true;
getSourcePosition.doNotCapture = true;
getFunctionScopeDetails.doNotCapture = true;
getFunctionScopeCount.doNotCapture = true;
// `GetFunctionScopeDetails` returns a raw JavaScript array. This enum enumerates the objects that
// are at specific indices of the array. We only care about one of these.
var V8ScopeDetailsFields;
(function (V8ScopeDetailsFields) {
    V8ScopeDetailsFields[V8ScopeDetailsFields["kScopeDetailsTypeIndex"] = 0] = "kScopeDetailsTypeIndex";
    V8ScopeDetailsFields[V8ScopeDetailsFields["kScopeDetailsObjectIndex"] = 1] = "kScopeDetailsObjectIndex";
    V8ScopeDetailsFields[V8ScopeDetailsFields["kScopeDetailsNameIndex"] = 2] = "kScopeDetailsNameIndex";
    V8ScopeDetailsFields[V8ScopeDetailsFields["kScopeDetailsStartPositionIndex"] = 3] = "kScopeDetailsStartPositionIndex";
    V8ScopeDetailsFields[V8ScopeDetailsFields["kScopeDetailsEndPositionIndex"] = 4] = "kScopeDetailsEndPositionIndex";
    V8ScopeDetailsFields[V8ScopeDetailsFields["kScopeDetailsFunctionIndex"] = 5] = "kScopeDetailsFunctionIndex";
})(V8ScopeDetailsFields || (V8ScopeDetailsFields = {}));
// getScopeForFunction extracts a V8ScopeDetails for the index'th element in the scope chain for the
// given function.
function getScopeForFunction(func, index) {
    const scopeDetails = getFunctionScopeDetails(func, index);
    return {
        scopeObject: scopeDetails[V8ScopeDetailsFields.kScopeDetailsObjectIndex],
    };
}
/**
 * Given a function and a free variable name, lookupCapturedVariableValue looks up the value of that free variable
 * in the scope chain of the provided function. If the free variable is not found, `throwOnFailure` indicates
 * whether or not this function should throw or return `undefined.
 *
 * @param func The function whose scope chain is to be analyzed
 * @param freeVariable The name of the free variable to inspect
 * @param throwOnFailure If true, throws if the free variable can't be found.
 * @returns The value of the free variable. If `throwOnFailure` is false, returns `undefined` if not found.
 */
function lookupCapturedVariableValue(func, freeVariable, throwOnFailure) {
    // The implementation of this function is now very straightforward since the intrinsics do all of the
    // difficult work.
    const count = getFunctionScopeCount(func);
    for (let i = 0; i < count; i++) {
        const scope = getScopeForFunction(func, i);
        if (freeVariable in scope.scopeObject) {
            return scope.scopeObject[freeVariable];
        }
    }
    if (throwOnFailure) {
        throw new Error("Unexpected missing variable in closure environment: " + freeVariable);
    }
    return undefined;
}
exports.lookupCapturedVariableValue = lookupCapturedVariableValue;
/**
 * Given a function, returns the name of the file where this function was defined.
 * Returns the empty string if the given function has no Script. (e.g. a native function)
 */
function getFunctionFile(func) {
    const script = getScript(func);
    return script ? script.name : "";
}
exports.getFunctionFile = getFunctionFile;
/**
 * Given a function, returns the line and column number in the file where this function was defined.
 * Returns { 0, 0 } if the location cannot be found or if the given function has no Script.
 */
function getFunctionLocation(func) {
    const script = getScript(func);
    if (script) {
        const pos = getSourcePosition(func);
        try {
            if (isNodeV10) {
                return scriptPositionInfo(script, pos);
            }
            else {
                return script.locationFromPosition(pos);
            }
        }
        catch (err) {
            // Be resilient to native functions not being available. In this case, we just return
            // '0,0'.  That's not great, but it at least lets us run, and it isn't a terrible
            // experience.
            //
            // Specifically, we only need these locations when we're printing out an error about not
            // being able to serialize something.  In that case, we still print out the names of the
            // functions (as well as the call-tree that got us there), *and* we print out the body
            // of the function.  With both of these, it is generally not too difficult to find out
            // where the code actually lives.
        }
    }
    return { line: 0, column: 0 };
}
exports.getFunctionLocation = getFunctionLocation;
